pygeos/_geom.pyx

```diff
@@ -0,0 +1,81 @@
+from cpython cimport PyObject
+cimport cython
+
+import numpy as np
+cimport numpy as np
+import pygeos
+
+from pygeos._geos cimport (
+    GEOSContextHandle_t,
+    GEOSGeometry,
+    GEOSGeom_clone_r,
+    GEOSGetGeometryN_r,
+    GEOS_init_r,
+    GEOS_finish_r,
+    get_geos_handle
+)
+from pygeos._pygeos_api cimport (
+    import_pygeos_c_api,
+    PyGEOS_CreateGeometry,
+    PyGEOS_GetGEOSGeometry
+)
+
+# initialize PyGEOS C API
+import_pygeos_c_api()
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def get_parts(object[:] array):
+    cdef Py_ssize_t geom_idx = 0
+    cdef Py_ssize_t part_idx = 0
+    cdef Py_ssize_t idx = 0
+    cdef GEOSGeometry *geom = NULL
+    cdef GEOSGeometry *part = NULL
+
+    cdef const np.intp_t [:] input_index_view = np.arange(0, len(array), dtype=np.intp)
+
+    counts = pygeos.get_num_geometries(array)
+    cdef int [:] counts_view = counts[:]
+
+    cdef Py_ssize_t count = counts.sum()
+
+    if count <= 0:
+        # return immeidately if there are no geometries to return
+        # count is negative when the only entries in array are None
+        return (
+            np.empty(shape=(0, ), dtype=np.object),
+            np.empty(shape=(0, ), dtype=np.intp)
+        )
+
+    parts = np.empty(shape=(count, ), dtype=np.object)
+    index = np.empty(shape=(count, ), dtype=np.intp)
+
+    cdef object[:] parts_view = parts[:]
+    cdef np.intp_t [:] index_view = index[:]
+
+    with get_geos_handle() as geos_handle:
```

Comment:
This uses the context manager approach from geopandas Cython vectorized branch.

Note that it does not add any handling of GEOS error / notice messages.  Added #217 to deal with this in a later PR unless it is essential we handle it here.

Reply:
Sounds good to handle it in a follow-up